File: app/src/main/java/com/rubyhuntersky/peregrine/AssetPrice.java
package com.rubyhuntersky.peregrine;

import android.os.Parcel;
import android.os.Parcelable;

import java.math.BigDecimal;

/**
 * @author wehjin
 * @since 1/29/16.
 */
public class AssetPrice implements Parcelable {
    public final String name;
    public final BigDecimal price;

    public AssetPrice(String name, BigDecimal price) {
        this.name = name;
        this.price = price;
    }

    public AssetPrice(Parcel in) {
        this(in.readString(), (BigDecimal) in.readSerializable());
    }

    public AssetPrice() {
        this("-", BigDecimal.ONE);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(name);
        dest.writeSerializable(price);
    }

    public static final Creator<AssetPrice> CREATOR = new Creator<AssetPrice>() {
        public AssetPrice createFromParcel(Parcel in) {
            return new AssetPrice(in);
        }

        public AssetPrice[] newArray(int size) {
            return new AssetPrice[size];
        }
    };
}

File: app/src/main/java/com/rubyhuntersky/peregrine/model/AssetPrice.kt
package com.rubyhuntersky.peregrine.model

import android.os.Parcel
import com.rubyhuntersky.peregrine.utility.DefaultParcelable
import com.rubyhuntersky.peregrine.utility.DefaultParcelable.Companion.generateCreator
import com.rubyhuntersky.peregrine.utility.read
import com.rubyhuntersky.peregrine.utility.toLabelAndCurrencyDisplayString
import com.rubyhuntersky.peregrine.utility.write
import java.math.BigDecimal

class AssetPrice @JvmOverloads constructor(val name: String = "-", val price: BigDecimal = BigDecimal.ONE) : DefaultParcelable {

    fun toDisplayString(): String = price.toLabelAndCurrencyDisplayString(name)

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.write(name, price)
    }

    companion object {
        @Suppress("unused")
        @JvmStatic val CREATOR = generateCreator { AssetPrice(it.read(), it.read()) }
    }
}
